### DISCUSSION TITLE
MOOSE WebControl doesn't work as expected

### DISCUSSION THREAD

**Discussion by huaianking** [idaholab/moose#28071] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hi MOOSE WebControl Developer, 

I tried to use MOOSE WebControl to increase my input Temperature at each time step using ex02 in the Thermal Hydraulics Module's tutorial. The control input file is listed below. I was able to run it without any error messages, but it doesn't get the job done. The T input value at each time step remains unchanged. 

I know there are other ways I can use to change T at each time step, but my main goal is to get webServerControl working so I can do coupling with other applications. 

I checked the source file and T in the InletMassFlowRateTemperature1Phase is already set as controllable. 

Can someone point out what I might have done wrong? Much appreciated! 
```
from MooseControl import MooseControl
command = '../../thermal_hydraulics-opt -i 02_coreWebControl.i'
command_list = command.split()
control = MooseControl(moose_command=command_list, moose_control_name='web_server')
control.initialize()
num_steps = 10
for t in range(num_steps):
    control.wait('TIMESTEP_BEGIN')
    pp_value = control.getPostprocessor('core_T_in')
    control.setContinue()
    control.wait('TIMESTEP_END')
    control.setControllableReal('Components/inlet/T',pp_value+100)
    control.setContinue()
control.finalize()
```

![image](https://github.com/idaholab/moose/assets/68820004/856c6db8-f096-43ab-8ea3-54cf5b674868)

**Comment by loganharbour** [comment]
~~Can you share the output from the `MooseControl`?~~

This isn't necessary.

**Comment by loganharbour** [comment]
@joshuahansel can you comment on controlling values with the `InletMassFlowRateTemperature1Phase`? `T` is definitely controllable but isn't the value of `T` used in the underlying `ADBoundaryFlux3EqnGhostMassFlowRateTemperature` object.

How do you guys deal with this kind of control with THM?

@huaianking the problem is that `T` is actually not controllable within that object.

**Reply by joshuahansel** [reply, marked as answer]
`T` should be controllable. It gets associated with `ADBoundaryFlux3EqnGhostMassFlowRateTemperature` via `connectObject(params, _boundary_uo_name, "T")`. We have a test for controlling this parameter:
`moose/modules/thermal_hydraulics/test/tests/components/inlet_mass_flow_rate_1phase/clg.ctrl_T_3eqn_rdg.i`.

**Reply by huaianking** [reply]
Thanks for the quick reply! Did I do something wrong here when I set up my input file? The T_in still doesn't seem to be controlled by the webControlServer when I run it for the testing case. But I run the test case you listed and it works just fine. 

```
T_in = 300. # K
m_dot_in = 1e-2 # kg/s
press = 10e5 # Pa

# core parameters
core_length = 1. # m
core_n_elems = 25
core_dia = '${units 2. cm -> m}'
core_pitch = '${units 8.7 cm -> m}'
A_core = '${fparse core_pitch^2 - 0.25 *pi * core_dia^2}'
P_wet_core = '${fparse 4*core_pitch + pi * core_dia}'
Dh_core = '${fparse 4 * A_core / P_wet_core}'

tot_power = 2000 # W

[GlobalParams]
  initial_p = ${press}
  initial_vel = 0.0001
  initial_T = ${T_in}
  gravity_vector = '0 0 0'

  rdg_slope_reconstruction = minmod
  scaling_factor_1phase = '1 1e-2 1e-4'
  closures = thm_closures
  fp = he
[]


[FluidProperties]
  [he]
    type = IdealGasFluidProperties
    molar_mass = 4e-3
    gamma = 1.67
    k = 0.2556
    mu = 3.22639e-5
  []
[]


[Closures]
  [thm_closures]
    type = Closures1PhaseTHM
  []
[]

[SolidProperties]
  [steel]
    type = ThermalFunctionSolidProperties
    rho = 8050
    k = 45
    cp = 466
  []
[]

[Components]
  [total_power]
    type = TotalPower
    power = ${tot_power}
  []
  [inlet]
    type = InletMassFlowRateTemperature1Phase
    input = 'core_chan:in'
    m_dot = ${m_dot_in}
    T = ${T_in}
  []

  [core_chan]
    type = FlowChannel1Phase
    position = '0 0 0'
    orientation = '0 0 1'
    length = ${core_length}
    n_elems = ${core_n_elems}
    roughness = .0001
    A = '${A_core}'
    D_h = ${Dh_core}
  []

  [core_hs]
    type = HeatStructureCylindrical
    position = '0 0 0'
    orientation = '0 0 1'
    length = ${core_length}
    n_elems = ${core_n_elems}
    names = 'block'
    widths = '${fparse core_dia / 2.}'
    solid_properties = 'steel'
    solid_properties_T_ref = '300'
    n_part_elems = 3
  []

  [core_heating]
    type = HeatSourceFromTotalPower
    hs = core_hs
    regions = block
    power = total_power
  []

  [core_ht]
    type = HeatTransferFromHeatStructure1Phase
    flow_channel = core_chan
    hs = core_hs
    hs_side = outer
    P_hf = '${fparse pi * core_dia}'
  []

  [outlet]
    type = Outlet1Phase
    input = 'core_chan:out'
    p = ${press}
  []
[]

[Postprocessors]
  [power_to_coolant]
    type = ADHeatRateConvection1Phase
    block = core_chan
    P_hf = '${fparse pi *core_dia}'
  []
  
  [core_T_in]
    type = SideAverageValue
    boundary = core_chan:in
    variable = T
  []

  [core_T_out]
    type = SideAverageValue
    boundary = core_chan:out
    variable = T
  []

  [core_p_in]
    type = SideAverageValue
    boundary = core_chan:in
    variable = p
  []

  [core_p_out]
    type = SideAverageValue
    boundary = core_chan:out
    variable = p
  []

  [core_delta_p]
    type = ParsedPostprocessor
    pp_names = 'core_p_in core_p_out'
    expression = 'core_p_in - core_p_out'
  []
[]

[Preconditioning]
  [pc]
    type = SMP
    full = true
  []
[]

[Controls]
  [web_server]
  type = WebServerControl
  execute_on='TIMESTEP_BEGIN TIMESTEP_END'
  []
[]

[Executioner]
  type=Transient
  solve_type = 'PJFNK'
  dt=10
  num_steps = 10
[]

#[Executioner]
 # type = Transient
  #start_time = 0

  #[TimeStepper]
   # type = IterationAdaptiveDT
   # dt = 10
  #[]
 # end_time = 5000

  #line_search = basic
  #solve_type = NEWTON

  #petsc_options_iname = '-pc_type'
  #petsc_options_value = 'lu'

  #nl_rel_tol = 1e-8
  #nl_abs_tol = 1e-8
  #nl_max_its = 25

#[]

[Outputs]
  csv=true
[]
```

**Reply by loganharbour** [reply]
Can you share the output that you get when running the python script that runs the `MooseControl`?
